src/ncp/ncp_base_mtd.cpp

```diff
@@ -2840,6 +2873,33 @@ void NcpBase::HandleEnergyScanResult(otEnergyScanResult *aResult)
     }
 }
 
+void NcpBase::HandleJoinerCallback_Jump(otError aError, void *aContext)
+{
+    static_cast<NcpBase *>(aContext)->HandleJoinerCallback(aError);
+}
+
+void NcpBase::HandleJoinerCallback(otError aError)
+{
+    otError error             = OT_ERROR_NONE;
+    bool    commission_status = (aError == OT_ERROR_NONE) ? true : false;
+
+    SuccessOrExit(error = mEncoder.BeginFrame(SPINEL_HEADER_FLAG | SPINEL_HEADER_IID_0, SPINEL_CMD_PROP_VALUE_IS,
+                                              SPINEL_PROP_THREAD_JOINER_COMMISSION));
+    SuccessOrExit(error = mEncoder.WriteBool(commission_status));
+    SuccessOrExit(error = mEncoder.EndFrame());
```

Comment:
How about if we emit a `VALUE_IS(LAST_STATUS)` with an `SPINEL_STATUS_JOIN_<>` to indicate success or failure of the join operation?

The errors from joiner callback are
```
/**
 * This function pointer is called to notify the completion of a join operation.
 *
 * @param[in]  aError    OT_ERROR_NONE if the join process succeeded.
 *                       OT_ERROR_SECURITY if the join process failed due to security credentials.
 *                       OT_ERROR_NOT_FOUND if no joinable network was discovered.
 *                       OT_ERROR_RESPONSE_TIMEOUT if a response timed out.
 * @param[in]  aContext  A pointer to application-specific context.
 *
 */
```

We already have `SPINEL_STATUS_JOIN_SECURITY` and `SPINEL_STATUS_JOIN_NO_PEERS` which map nicely to two of the error cases above.

The advantage of this model is we can use `mChangedPropSet` and this way the async event is guaranteed to emitted even if there is no buffer at the moment.

We can consider adding a `SPINEL_STATUS_JOIN_SUCCESS` and `SPINEL_STATUS_JOIN_TIMEOUT` to indicate the other two possible outcomes.

Thoughts?

Reply:
Good proposal.  Will update.